src/shared/Core/StringExtensions.cs

```diff
@@ -240,5 +241,15 @@ public static bool Contains(this string str, string value, StringComparison comp
         {
             return str?.IndexOf(value, comparisonType) >= 0;
         }
+
+        /// <summary>
+        /// Convert string to snake case.
+        /// </summary>
+        /// <param name="str">String to convert.</param>
+        /// <returns>Input string converted to snake case.</returns>
+        public static string ToSnakeCase(this string str) =>
+            string.Concat(str.Select((x, i) => (i > 0 && char.IsUpper(x) &&
+                (char.IsLower(str[i - 1]) || char.IsLower(str[i + 1])))
+                ? "_" + x : x.ToString())).ToLower();
```

Comment:
This was a little hard to read 😵‍💫

But I think there's also a bug here. Try using "OO" as an input.. we should get "oo" but we get an exception.

The problem is we are doing `str[i+1]` without checking that `i < str.Length - 1`.

Reply:
I wondered if we could use `StringBuilder` here and avoid LINQ to improve readability and improve perf.

I used BenchmarkDotnet to compare an (expanded out) LINQ version (with the `i < str.Length - 1` check) and a `StringBuilder + for-loop` version; I found the latter to be 42% of the compute and 22% of the memory of the former.

```csharp
[MemoryDiagnoser]
public class SnakeCaseBenchmark
{
    private const string str = "oFoObArFoObAzF";

    [Benchmark]
    public string Builder()
    {
        int len = str.Length;
        var sb = new StringBuilder(2*len);
        for (int i = 0; i < len; i++)
        {
            if (i > 0 && char.IsUpper(str[i]) &&
                (char.IsLower(str[i - 1]) || i < len - 1 && char.IsLower(str[i + 1])))
            {
                sb.Append('_');
            }

            sb.Append(char.ToLower(str[i]));
        }

        return sb.ToString();
    }

    [Benchmark(Baseline = true)]
    public string Linq()
    {
        int len = str.Length;
        var x = str.Select((x, i) =>
        {
            if (i > 0 && char.IsUpper(x) && (char.IsLower(str[i - 1]) || i < len - 1 && char.IsLower(str[i + 1])))
            {
                return "_" + x;
            }

            return x.ToString();
        });

        return string.Concat(x).ToLower();
    }
}
```


```text
BenchmarkDotNet=v0.13.5, OS=macOS Ventura 13.2.1 (22D68) [Darwin 22.3.0]
Apple M1 Max, 1 CPU, 10 logical and 10 physical cores
.NET SDK=8.0.100-preview.1.23115.2
  [Host]     : .NET 6.0.13 (6.0.1322.58009), Arm64 RyuJIT AdvSIMD
  DefaultJob : .NET 6.0.13 (6.0.1322.58009), Arm64 RyuJIT AdvSIMD


|  Method |     Mean |   Error |  StdDev | Ratio |   Gen0 | Allocated | Alloc Ratio |
|-------- |---------:|--------:|--------:|------:|-------:|----------:|------------:|
| Builder | 165.6 ns | 0.47 ns | 0.42 ns |  0.42 | 0.0918 |     192 B |        0.22 |
|    Linq | 390.6 ns | 0.78 ns | 0.73 ns |  1.00 | 0.4244 |     888 B |        1.00 |
```

Reply:
Great analysis! I am more than happy to switch to Builder.